packages/tools/src/utilities/segmentation/contourAndFindLargestBidirectional.ts

```diff
@@ -0,0 +1,294 @@
+  const { sliceContours } = contours;
+  const { segmentIndex, containedSegmentIndices } = segment;
+  let maxBidirectional;
+  const isInSegment = createIsInSegment(
+    segVolumeId,
+    segmentIndex,
+    containedSegmentIndices
+  );
+  for (const sliceContour of sliceContours) {
+    const bidirectional = createBidirectionalForSlice(
+      sliceContour,
+      isInSegment,
+      maxBidirectional
+    );
+    if (!bidirectional) {
+      continue;
+    }
+    maxBidirectional = bidirectional;
+  }
+  if (maxBidirectional) {
+    Object.assign(maxBidirectional, segment);
+  }
+  return maxBidirectional;
+}
+
+const lengthSq = (point) =>
+  point[0] * point[0] + point[1] * point[1] + point[2] * point[2];
+
+const distanceSq = (p1, p2) => lengthSq(vec3.sub(vec3.create(), p1, p2));
+
+/**
+ * Determines if there is a point between point1 and point2 which is not
+ * contained in the segmentation
+ */
+function isLineInSegment(
+  point1: Types.Point3,
+  point2: Types.Point3,
+  isInSegment
+) {
+  const ijk1 = isInSegment.toIJK(point1);
+  const ijk2 = isInSegment.toIJK(point2);
+  const testPoint = vec3.create();
+  const { testIJK } = isInSegment;
+  const delta = vec3.sub(vec3.create(), ijk1, ijk2);
+
+  // Test once for index value between the two points, so the max of the
+  // difference in IJK values
+  const testSize = Math.round(Math.max(...delta.map(Math.abs)));
+  if (testSize < 2) {
+    // No need to test when there are only two points
+    return true;
+  }
+  const unitDelta = vec3.scale(vec3.create(), delta, 1 / testSize);
+
+  for (let i = 1; i < testSize; i++) {
+    vec3.scaleAndAdd(testPoint, ijk2, unitDelta, i);
+    if (!testIJK(testPoint)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ * This function creates a bidirectional data object for the given slice and
+ * slice contour, only when the major distance is larger than currentMax, or
+ * equal to current max and the minor is larger than currentMax's minor.
+ * It does this by looking at every pair of distances in sliceCountour to find
+ * those larger than the currentMax, and then finds the minor distance for those
+ * major distances.
+ *
+ */
+function createBidirectionalForSlice(
+  sliceContour,
+  isInSegment,
+  currentMax = { maxMajor: 0, maxMinor: 0 }
+) {
+  const { points } = sliceContour.polyData;
+  const { maxMinor: currentMaxMinor, maxMajor: currentMaxMajor } = currentMax;
+  let maxMajor = currentMaxMajor * currentMaxMajor;
+  let maxMinor = currentMaxMinor * currentMaxMinor;
+  let maxMajorPoints;
+  for (let index1 = 0; index1 < points.length; index1++) {
+    for (let index2 = index1 + 1; index2 < points.length; index2++) {
+      const point1 = points[index1];
+      const point2 = points[index2];
+      const distance2 = distanceSq(point1, point2);
+      if (distance2 < maxMajor) {
+        continue;
+      }
+      if (distance2 === maxMajor && maxMajorPoints) {
```

Comment:
this should use isEqual to include EPS

Reply:
THat would need several more multiplies since it would need to use a multiplicative epsilon.  I agree ideally it would include EPS, but I don't think the performance hit is worthwhile.  I can add a simple + EPSILON in a few places if you want - that would get most of the relevant differences.